From 4e9f1d9789eccdcffaa42ada2c87b52890f5abe8 Mon Sep 17 00:00:00 2001
From: Steve Ellcey <sellcey@caviumnetworks.com> 
Date: Sun, 27 May 2018 15:08:41 +0000
Subject: [PATCH 2/2] Enable ILP32 mode in gdb on aarch64.

Signed-off-by: Steve Ellcey <sellcey@caviumnetworks.com>

[Patch v2 2/2] Enable ILP32 mode in gdb on aarch64
Here is the second of two gdb patches to support aarch64 ILP32 mode.
This one just has the bfd changes. Along with changes to cpu-aarch64.c
and elfnn-aarch64.c parts before, this modification to elfxx-
aarch64.c to fix gdb support in ILP32 mode.

        * elfxx-aarch64.c (_bfd_aarch64_elf_grok_prstatus): Handle
        size and offset of ILP32 executables.
        (_bfd_aarch64_elf_grok_psinfo):Ditto.

Signed-off-by: Manjunath Matti <mmatti@mvista.com>

Upstream-Status: Submitted [https://sourceware.org/ml/binutils/2017-03/msg00075.html]
---
 bfd/elfxx-aarch64.c | 29 ++++++++++++++++++++++++-----
 1 file changed, 24 insertions(+), 5 deletions(-)

diff --git a/bfd/elfxx-aarch64.c b/bfd/elfxx-aarch64.c
index 9d4aa8fb20..c2cad140e6 100644
--- a/bfd/elfxx-aarch64.c
+++ b/bfd/elfxx-aarch64.c
@@ -561,14 +561,25 @@ _bfd_aarch64_elf_grok_prstatus (bfd *abfd, Elf_Internal_Note *note)
       default:
 	return FALSE;
 
-      case 392:		/* sizeof(struct elf_prstatus) on Linux/arm64.  */
+      case 352:	/* sizeof(struct elf_prstatus) on Linux/aarch64 ilp32.  */
 	/* pr_cursig */
-	elf_tdata (abfd)->core->signal
-	  = bfd_get_16 (abfd, note->descdata + 12);
+	elf_tdata (abfd)->core->signal = bfd_get_16 (abfd, note->descdata + 12);
 
 	/* pr_pid */
-	elf_tdata (abfd)->core->lwpid
-	  = bfd_get_32 (abfd, note->descdata + 32);
+	elf_tdata (abfd)->core->lwpid = bfd_get_32 (abfd, note->descdata + 32);
+
+	/* pr_reg */
+	offset = 72;
+	size = 272;
+
+	break;
+
+      case 392:	/* sizeof(struct elf_prstatus) on Linux/aarch64.  */
+	/* pr_cursig */
+	elf_tdata (abfd)->core->signal = bfd_get_16 (abfd, note->descdata + 12);
+
+	/* pr_pid */
+	elf_tdata (abfd)->core->lwpid = bfd_get_32 (abfd, note->descdata + 32);
 
 	/* pr_reg */
 	offset = 112;
@@ -589,6 +600,13 @@ _bfd_aarch64_elf_grok_psinfo (bfd *abfd, Elf_Internal_Note *note)
     {
     default:
       return FALSE;
+    case 124:         /* sizeof(struct elf_prpsinfo) on Linux/aarch64 ilp32.  */
+      elf_tdata (abfd)->core->pid = bfd_get_32 (abfd, note->descdata + 12);
+      elf_tdata (abfd)->core->program
+	= _bfd_elfcore_strndup (abfd, note->descdata + 28, 16);
+      elf_tdata (abfd)->core->command
+	= _bfd_elfcore_strndup (abfd, note->descdata + 44, 80);
+      break;
 
     case 136:	     /* This is sizeof(struct elf_prpsinfo) on Linux/aarch64.  */
       elf_tdata (abfd)->core->pid = bfd_get_32 (abfd, note->descdata + 24);
@@ -596,6 +614,7 @@ _bfd_aarch64_elf_grok_psinfo (bfd *abfd, Elf_Internal_Note *note)
 	= _bfd_elfcore_strndup (abfd, note->descdata + 40, 16);
       elf_tdata (abfd)->core->command
 	= _bfd_elfcore_strndup (abfd, note->descdata + 56, 80);
+      break;
     }
 
   /* Note that for some reason, a spurious space is tacked
-- 
2.11.1

